_api/hasura/cron/syncCoSouls.ts

```diff
@@ -216,24 +229,118 @@ const syncCoSoulToken = async (
+      );
+    } else {
+      console.log(
+        'skipping setting on-chain PGIVE because it is 0 for tokenId: ' +
+          cosoul.token_id +
+          ' address: ' +
+          cosoul.address
+      );
+    }
+  }
+  if (payload.length > 4) {
+    // 0x and one byte for slot 0x00
+    //if only one needs to be updated on chain, we can use setOnChainPGIVE
+    if (payload.length === 20) {
+      //0x00 plus 4 bytes for pgive and 4 bytes for tokenId
+      const tx = await setOnChainPGIVE(
+        updatedCosouls[0].cosoul.token_id,
+        updatedCosouls[0].localPGIVE
+      );
+      await tx.wait();
+    } else {
+      const tx = await setBatchOnChainPGIVE(payload);
+      await tx.wait();
+    }
+    successMessages.forEach(msg => console.log(msg));
+  }
+
+  await Promise.allSettled(
+    updatedCosouls.map(async ({ cosoul, localPGIVE }) => {
+      try {
+        await adminClient.mutate(
+          {
+            update_cosouls_by_pk: [
+              {
+                pk_columns: {
+                  id: cosoul.id,
+                },
+                _set: {
+                  pgive: localPGIVE,
+                  checked_at: new Date().toISOString(),
+                  synced_at: new Date().toISOString(),
+                },
+              },
+              {
+                id: true,
+              },
+            ],
+          },
+          {
+            operationName: 'cron__updateCoSoulCache',
+          }
+        );
+        console.log(
+          'Updated CoSoul with address: ' +
+            cosoul.address +
+            ' to tokenId: ' +
+            cosoul.token_id
+        );
+      } catch (e: unknown) {
+        const message =
+          'Error updating Cosoul with addesss ' +
+          cosoul.address +
+          ' to tokenId: ' +
+          cosoul.token_id;
+        console.error(message);
+        if (e instanceof Error) throw new Error(message);
+      }
+      return;
+    })
+  );
+};
+
+async function updateCoSoulOnChain(
+  updatedCosouls: { cosoul: CoSoul; localPGIVE: number }[]
+) {
   try {
-    await syncCoSoulToken(
-      cosoul.id,
-      cosoul.address,
-      totalPGIVE,
-      cosoul.token_id
-    );
+    if (updatedCosouls.length === 1) {
+      await syncCoSoulToken(
+        updatedCosouls[0].cosoul.id,
+        updatedCosouls[0].cosoul.address,
+        updatedCosouls[0].localPGIVE,
+        updatedCosouls[0].cosoul.token_id
+      );
+    } else {
+      await syncPatchCoSoulToken(updatedCosouls);
+    }
     return true;
   } catch (e: any) {
     // don't ruin the whole thing, this might be an on-chain issue or temporary setback
     // TODO: send this to sentry
+    const failedBatch = updatedCosouls.reduce(
+      (accumulator, currentValue, currentIndex) =>
+        accumulator +
+        `[Cosoul ${currentIndex} with id: ${currentValue.cosoul.id}, tokenId: ${currentValue.cosoul.token_id}, address: 
+      ${currentValue.cosoul.address} ,and targetPIVE: ${currentValue.localPGIVE}]`,
```

Comment:
the ` ${currentValue.cosoul.address} ,and` seems weird, you mean `${currentValue.cosoul.address}, and` ?